sotodlib/hwp/g3thwp.py

```diff
@@ -621,7 +629,134 @@ def write_solution(self, solved, output=None):
+            self._write_empty_solution_h5(tod, output, h5_address)
+            return
+            
+        if len(data) == 0:
+            logger.warning('No HWP data in the specified timestamps.')
+            self._write_empty_solution_h5(tod, output, h5_address)
+            return
+        
+        # calculate HWP angle 
+        logger.debug("analyze")
+        
+        try:
+            solved = self.analyze(data)
+        except Exception as e:
+            logger.error(f"Exception '{e}' thrown while calculating HWP angle. Encoder signal might have too much noise.")
+            self._write_empty_solution_h5(tod, output, h5_address)
+            return 
+            
+        if len(solved['fast_time']) == 0:
+            logger.info('No rotation data in the specified timestamps.')
+            self._write_empty_solution_h5(tod, output, h5_address)
+            return
+            
+        # calculate template subtracted angle
+        try:
+            self.eval_angle(solved)
+        except Exception as e:
+            logger.error(f"Exception '{e}' thrown while the template subtraction")
+            pass
+        
+        # write solution
+        aman = sotodlib.core.AxisManager(tod.dets, tod.samps)
+        aman.wrap_new('timestamps', shape=('samps', ))
+        aman.wrap_new('hwp_angle_ver1', shape=('samps', ))
+        aman.wrap_new('hwp_angle_ver2', shape=('samps', ))
+        aman.wrap_new('stable', shape=('samps', ))
+        aman.wrap_new('locked', shape=('samps', ))
+        aman.wrap_new('hwp_rate', shape=('samps', ))
+        aman.wrap_new('eval', shape=('samps', ))
+        
+        aman.timestamps[:] = tod.timestamps
+        aman.stable[:] = scipy.interpolate.interp1d(solved['slow_time'], solved['stable'], kind='linear', bounds_error=False)(tod.timestamps)
+        aman.locked[:] = scipy.interpolate.interp1d(solved['slow_time'], solved['locked'], kind='linear', bounds_error=False)(tod.timestamps)
+        aman.hwp_rate[:] = scipy.interpolate.interp1d(solved['slow_time'], solved['hwp_rate'], kind='linear', bounds_error=False)(tod.timestamps)
+        if 'fast_time_raw' in solved.keys():
+            aman.hwp_angle_ver1[:] = scipy.interpolate.interp1d(solved['fast_time_raw'], solved['angle'], kind='linear',bounds_error=False)(tod.timestamps)
+            aman.hwp_angle_ver2[:] = scipy.interpolate.interp1d(solved['fast_time'], solved['angle'], kind='linear',bounds_error=False)(tod.timestamps)
+            aman.eval[:] = np.ones(len(tod.timestamps))
+        else:
+            logger.info('Template subtraction failed')
+            aman.hwp_angle_ver1[:] = scipy.interpolate.interp1d(solved['fast_time'], solved['angle'], kind='linear', bounds_error=False)(tod.timestamps)
+            aman.hwp_angle_ver2[:] = scipy.interpolate.interp1d(solved['fast_time'], solved['angle'], kind='linear', bounds_error=False)(tod.timestamps)
```

Comment:
You have to do something to handle the discontinuities in the angle due to the periodic domain (0 - 360).

numpy.unwrap can be helpful here.

There are probably other places in sotodlib where there has been done.

Reply:
Fixed. We can select the output type (continuous/discontinuous) of the angle solution with mod2pi argument in "analyze" function.

Reply:
I'm not sure I understand... the call to analyze() is still in "discontinuous" mode, which means you're interpolating a curve with discontinuities.

This is what I'm worried about -- the black points are your solution points (fast_time, angle) and the blue points are tod.timestamps.  The interpolation is not valid at the points where angle wraps from 2pi to 0 (or 0 to 2pi).

![image](https://github.com/simonsobs/sotodlib/assets/1054988/01321976-208b-4b1c-a8cd-d3903f45980a)

Reply:
Yes, discontinuous output of the angle can arise the undesired blue points.
But in my environment, the angle solution is now output in the continuous mode with the call:
`solved = self.analyze(data, mod2pi=False)`,
on L745 of g3thwp.py.
I'm sorry but I would like you to try to run the script again.
![image](https://github.com/simonsobs/sotodlib/assets/75608221/e4b9ded9-4ec1-4535-b732-d4f49e37c7df)